### PR TITLE
HW2-HW3

### PR DESCRIPTION
**Pull request by DataJoshua**
### Summary

A brief description of the pull request.

### How it works

Screenshots/screencasts of the pull request introduced functionality.

### Test plan

List of steps to manually test introduced functionality:

- Log in
- Go to http://localhost:3000

### Deploy notes

Notes regarding deployment the contained body of work.
These should note any db migrations, services, scripts, etc.

### Checklist:

- [ ] I have performed a self-review of my own code
- [ ] My pull request is shorter than 500 lines

### CODE REVIEW

**Comment by DataJoshua** [comment]
Д3 сделал в branch

File: app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      @project = Project.find(params[:project_id])
+    end
+  
+    def set_task
+      @task = Task.find(params[:id])
+    end
+
+    def task_params
+      params.require(:task).permit(:name, :title, :description, :deadline_at)
+    end
```
Comment (RasimKhusaenov) [inline app/controllers/tasks_controller.rb:64]: Why you duplicate task_params?

File: app/controllers/tasks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  
+    private
+
+    def task_params
+      params.require(:task).premit(:title, :name, :description)
+    end
+
+    def set_project
+      @project = Project.find(params[:project_id])
```
Comment (RasimKhusaenov) [inline app/controllers/tasks_controller.rb:55]: It is okay, but in the assignment I asked to use bang version of `find_by` method

File: app/helpers/tasks_helper.rb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+module TasksHelper
+end
```
Comment (RasimKhusaenov) [inline app/helpers/tasks_helper.rb:2]: Redundant. We don't need it.

File: app/javascript/packs/application.js [hunk 4]
```diff
@@ -1,12 +1,18 @@
+/* eslint no-console:0 */
```
Comment (RasimKhusaenov) [inline app/javascript/packs/application.js:1]: Don't edit this file

File: app/views/projects/index.html.erb [hunk 5]
```diff
@@ -18,7 +18,7 @@
         <td><%= project.description %></td>
         <td><%= link_to 'Show', project %></td>
         <td><%= link_to 'Edit', edit_project_path(project) %></td>
-        <td><%= link_to 'Destroy', project, method: :delete, data: { confirm: 'Are you sure?' } %></td>
+        <td><%= link_to 'Destroy', project, method: :destroy, data: { confirm: 'Are you sure?' } %></td>
```
Comment (RasimKhusaenov) [inline app/views/projects/index.html.erb:21]: Does it work?

File: app/views/tasks/_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        <%= form.label :title %>
+        <%= form.text_field :title %>
+    </div>
+    <div>
+        <%= form.label :description %>
+        <%= form.text_field :description %>
+    </div>
+    <div>
+        <%= form.label :deadline_at %>
+        <input type="date" name="task[deadline_at]" id="task[deadline_at]" value=<%= Time.now + 7.days %> />
```
Comment (RasimKhusaenov) [inline app/views/tasks/_form.html.erb:17]: It is incorrect. In this implementation, the user can change the default deadline value

File: app/views/tasks/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+    <% @tasks.each do |task| %>
+        <li>
+            <%= link_to task.name, url_for([@project, task]) %>
+        </li>
+     
+    <% end %>
+</ul>
+
+<%= link_to "Create Task", new_project_task_path %>
+<%= link_to "Back", @project %>
```
Comment (RasimKhusaenov) [inline app/views/tasks/index.html.erb:13]: Incorrect link. We should back to project tasks

File: app/views/tasks/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+<ul>
+    <% @tasks.each do |task| %>
+        <li>
+            <%= link_to task.name, url_for([@project, task]) %>
+        </li>
+     
+    <% end %>
+</ul>
+
+<%= link_to "Create Task", new_project_task_path %>
```
Comment (RasimKhusaenov) [inline app/views/tasks/index.html.erb:12]: Does it work?

File: app/views/tasks/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<h2><%= @task.name %></h2>
+<p><%= @task.title %></p>
+<p><%= @task.description %></p>
+<p>Deadline: <%= @task.deadline_at %></p>
+
+<div>
+    <%= link_to 'back', project_tasks_path %>
```
Comment (RasimKhusaenov) [inline app/views/tasks/show.html.erb:7]: Does it work?

File: config/database.yml [hunk 10]
```diff
@@ -22,8 +22,13 @@ default: &default
   pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
 
 development:
-  <<: *default
-  database: task_tracker_development
+  adapter: postgresql
```
Comment (RasimKhusaenov) [inline config/database.yml:25]: Don't edit this file

File: db/migrate/20220927132936_create_tasks.rb [hunk 11]
```diff
@@ -0,0 +1,12 @@
+class CreateTasks < ActiveRecord::Migration[6.1]
+  def change
+    create_table :tasks do |t|
+      t.string :project_id
+      t.string :name
+      t.string :title
+      t.string :description
+      t.datetime :deadline_at, :default => Time.now
```
Comment (RasimKhusaenov) [inline db/migrate/20220927132936_create_tasks.rb:8]: Remove default. It sets default value to date when migration was migrated

File: package.json [hunk 12]
```diff
@@ -5,6 +5,7 @@
     "@rails/activestorage": "^6.0.0",
     "@rails/ujs": "^6.0.0",
     "@rails/webpacker": "5.4.3",
+    "bundle": "^2.1.0",
```
Comment (RasimKhusaenov) [inline package.json:8]: Why?

File: app/models/task.rb [hunk 13]
```diff
@@ -0,0 +1,7 @@
+class Task < ApplicationRecord
+
+    validates :name, presence: true
```
Comment (RasimKhusaenov) [inline app/models/task.rb:3]: :+1:

File: app/views/tasks/index.html.erb [hunk 14]
```diff
@@ -0,0 +1,13 @@
+<h1>Tasks</h1>
+
+<ul>
+    <% @tasks.each do |task| %>
+        <li>
+            <%= link_to task.name, url_for([@project, task]) %>
+        </li>
+     
+    <% end %>
+</ul>
```
Comment (RasimKhusaenov) [inline app/views/tasks/index.html.erb:10]: There should be a table like the projects. With "Show", "Edit", "Destroy" buttons.

File: app/controllers/tasks_controller.rb [hunk 15]
```diff
@@ -0,0 +1,65 @@
+class TasksController < ApplicationController
+  before_action :set_project
```
Comment (RasimKhusaenov) [inline app/controllers/tasks_controller.rb:2]: Do you need it everywhere?

**Comment by RasimKhusaenov** [comment]
⚠️ You should do another HWs in another branch, hand it in another PR.

**Comment by DataJoshua** [comment]
okay

File: app/controllers/tasks_controller.rb [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    @task = @project.tasks.find(params[:id]) 
+  end 
+
+  def create
+    @task = @project.tasks.create(task_params)
+    redirect_to [@project, @task]
+  end
+
+  def destroy
+    @task = @project.tasks.find(:id)
```
Comment (RasimKhusaenov) [inline app/controllers/tasks_controller.rb:23]: Does it work?

File: app/controllers/tasks_controller.rb [hunk 17]
```diff
@@ -0,0 +1,59 @@
+class TasksController < ApplicationController
+  before_action :set_project
+  before_action :set_task, only: %i[show edit update destroy]
```
Comment (RasimKhusaenov) [inline app/controllers/tasks_controller.rb:3]: You don't use it anywhere, but you should use it almost everywhere

File: app/views/projects/show.html.erb [hunk 18]
```diff
@@ -12,3 +12,7 @@
 
 <%= link_to 'Edit', edit_project_path(@project) %> |
 <%= link_to 'Back', projects_path %>
+
+
+<%= link_to 'tasks', [@project, @project.tasks.build] %>
```
Comment (RasimKhusaenov) [inline app/views/projects/show.html.erb:17]: Does it work?

File: app/views/tasks/new.html.erb [hunk 19]
```diff
@@ -0,0 +1,4 @@
+
+<h1>Create<h1>
+
+<%= render "form", task: @project.tasks.build %>
```
Comment (RasimKhusaenov) [inline app/views/tasks/new.html.erb:4]: Use task from controller

File: db/migrate/20220927132936_create_tasks.rb [hunk 20]
```diff
@@ -0,0 +1,12 @@
+class CreateTasks < ActiveRecord::Migration[6.1]
+  def change
+    create_table :tasks do |t|
+      t.string :project_id
```
Comment (RasimKhusaenov) [inline db/migrate/20220927132936_create_tasks.rb:4]: Use reference, not string

File: app/views/tasks/index.html.erb [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+    <thead/>
+    <tbody>
+        <% @tasks.each do |task| %>
+            <tr>
+                <td> <%= task.name %> </td>
+                <td> <%= task.description %> </td>
+                <td> <%= link_to "Show", project_task_path(@project, task) %> </td>
+                <td> <%= link_to "Edit", edit_project_task_path(@project, task) %> </td> 
+                <td> <%= link_to "Destroy", [@project, task], data: {turbo_method: :delete}  %> </td>
+            </tr>
```
Comment (RasimKhusaenov) [inline app/views/tasks/index.html.erb:18]: Why you need turbo_method?

File: app/controllers/tasks_controller.rb [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+
+  def create
+    @task = @project.tasks.create(task_params)
+    redirect_to [@project, @task]
+  end
+
+  def destroy
+    @task = @project.tasks.find(:id)
+    
+    if @task.destroy
```
Comment (RasimKhusaenov) [inline app/controllers/tasks_controller.rb:25]: Why we need this condition?

**Review by RasimKhusaenov — CHANGES_REQUESTED** [review]

File: app/controllers/tasks_controller.rb [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+  def create
+    @task = @project.tasks.create(task_params)
+    redirect_to [@project, @task]
+  end
+
+  def destroy
+    @task = @project.tasks.find(params[:id])
+    
+    @task.destroy
+    redirect_to project_tasks_path, notice: "Project was successfully destroyed."
```
Comment (RasimKhusaenov) [inline app/controllers/tasks_controller.rb:26]: You should redirect to deleted task's project page

File: app/controllers/tasks_controller.rb [hunk 24]
```diff
@@ -0,0 +1,58 @@
+class TasksController < ApplicationController
+  before_action :set_project, only: %i[index new show create destroy edit update]
+  before_action :set_task, only: %i[show edit update destroy]
+
+  def index
+    @tasks = @project.tasks.all
```
Comment (RasimKhusaenov) [inline app/controllers/tasks_controller.rb:6]: `All` is redundant here
```suggestion
    @tasks = @project.tasks
```

File: app/controllers/tasks_controller.rb [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+  def new
+    @task = @project.tasks.build
+  end
+
+  def show
+    @task = @project.tasks.find(params[:id]) 
+  end 
+
+  def create
+    @task = @project.tasks.create(task_params)
```
Comment (RasimKhusaenov) [inline app/controllers/tasks_controller.rb:18]: What happens if I try to create a task with an empty name?

File: app/controllers/tasks_controller.rb [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+      redirect_to [@project, @task]
+    else
+      render :edit
+    end
+  end
+  
+  private
+
+  def task_params
+    params.require(:task).premit(:title, :name, :description)
```
Comment (RasimKhusaenov) [inline app/controllers/tasks_controller.rb:44]: Why you duplicate task_params?

File: app/controllers/tasks_controller.rb [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def set_project
+    @project = Project.find(params[:project_id])
```
Comment (RasimKhusaenov) [inline app/controllers/tasks_controller.rb:48]: It is okay, but in the assignment I asked to use bang version of find_by method

File: app/controllers/tasks_controller.rb [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+  end
+  
+  def set_task
+    @task = Task.find(params[:id])
```
Comment (RasimKhusaenov) [inline app/controllers/tasks_controller.rb:52]: It is okay, but in the assignment I asked to use bang version of find_by method

File: db/migrate/20221022093453_add_reference_to_tasks.rb [hunk 29]
```diff
@@ -0,0 +1,6 @@
+class AddReferenceToTasks < ActiveRecord::Migration[6.1]
+  def change
+    remove_column :tasks, :project_id
+    add_reference :tasks, :project, foreign_key: true
```
Comment (RasimKhusaenov) [inline db/migrate/20221022093453_add_reference_to_tasks.rb:4]: Not in new migration! You just added migration in this PR, it's not in the project yet. Therefore, you should change the existing one, and not produce new ones.

**Review by RasimKhusaenov — CHANGES_REQUESTED** [review]

**Pull request closed without merging**